### DISCUSSION TITLE
Can Partition and Partition Key exist simultaneously？

### DISCUSSION THREAD

**Discussion by CharlesWu123** [milvus-io/milvus#34569] [Q&A and General discussion] [answered]
1. I want to use both Partition and Partition Key to achieve two fast filtering methods simultaneously. For example, I often have two filtering conditions in my filtering criteria: market and plate, but I can only use one as the Partition Key. Can I use the other as the Partition name? I wonder if Milvus supports this setting.
2. If there are no filtering conditions when querying, will creating partitions or partition keys and not creating partitions or partition keys have a performance impact?

**Comment by yhmo** [comment]
Partition is to split data into different physical paths in storage.
There are two kinds of partitions:
1. manual partitions, created by collection.create_partiiton(), manually created by users, can be dropped. Users can create a new partition at any time. User must specify a partition to insert data, otherwise, the data will be inserted into the default partition. During search, user can specify a partition to search, otherwise, it will search in all the partitions.

2. partitions pre-defined by partition key. Once the collection is created with a partition key, the partitions are automatically created. User can change the num_partitions value to define how many partitions be created.
`collection = Collection(name="xxx", schema=schema, num_partitions=10)`
Once a partition key is defined, the collection doesn't allow to create or drop any partitions.
During search, the search engine can quickly determine which partition should be searched according to the filter expression(if the expression is like "partition_key_name == 4"). If the expression is not like the format, the search engine will scan all the partitions.


You cannot apply both manual partitions and partition key in the same collection.

**Reply by yhmo** [reply]
Performance impact:
If the search engine knows which partition should be searched, the search performance will be good.

So, for manual partitions, if you specify a partition for search(), the performance is good. Otherwise, it will search the entire collection.
For partition key, if your filter expression is like "partition_key_name == 4", the performance is good. Otherwise, it will search the entire collection.

**Reply by CharlesWu123** [reply]
Thanks, The second question I want to ask is: If I do not specify a filtering condition on the partition key, will there be a performance impact when using a partition key versus not using one?

**Reply by xiaofan-luan** [reply, marked as answer]
> Thanks, The second question I want to ask is: If I do not specify a filtering condition on the partition key, will there be a performance impact when using a partition key versus not using one?

it might or might not. 
if you only have 1M data, then split data into 64 partition or more definitely affect performance(because original you put all data in one segment but with partitions you need to search 64 segments).
On the other size, if data volume is huge then it doesn't really matter

**Reply by maxshatskiy** [reply]
> Performance impact: If the search engine knows which partition should be searched, the search performance will be good.
> 
> So, for manual partitions, if you specify a partition for search(), the performance is good. Otherwise, it will search the entire collection. For partition key, if your filter expression is like "partition_key_name == 4", the performance is good. Otherwise, it will search the entire collection.

Do I understand correctly, thatn if my filter is "partition_key_name IN [4,5,6]" or even "partition_key_name IN [4,5,6] and other_field=X", then search will still be done in all partitions in both cases, and having partition in this case does not have speed advantage in terms of search and querying?

**Reply by yhmo** [reply]
Partition key works both for "partition_key_name IN [4,5,6]" and "partition_key_name IN [4,5,6] and other_field=X", it only searches a few partitions for [4, 5, 6].

"partition_key_name IN [4,5,6]" --------------------------- searches a few partitions for [4, 5, 6]
"partition_key_name IN [4,5,6] and other_field=X" -------------- searches a few partitions for [4, 5, 6] and ignore the data in these partitions that  other_field != X.
